src/StateService/StatePlugin.cs

```diff
@@ -233,5 +235,81 @@ public JObject GetStateHeight(JArray _params)
             json["validatedrootindex"] = StateStore.Singleton.ValidatedRootIndex;
             return json;
         }
+
+        private ContractState GetHistoricalContractState(MPTTrie<StorageKey, StorageItem> trie, UInt160 script_hash)
+        {
+            const byte prefix = 8;
+            StorageKey skey = new KeyBuilder(NativeContract.ContractManagement.Id, prefix).Add(script_hash);
+            return trie[skey]?.GetInteroperable<ContractState>();
+        }
+
+        private void PrepareStateParams(JArray _params, out MPTTrie<StorageKey, StorageItem> trie, out StorageKey skey)
+        {
+            if (!uint.TryParse(_params[0].AsString(), out uint index))
+            {
+                if (!UInt256.TryParse(_params[0].AsString(), out UInt256 root_hash))
+                    throw new RpcException(-100, "Invalid block hash or index");
+                var header = NativeContract.Ledger.GetHeader(System.StoreView, root_hash);
+                if (header is null)
+                    throw new RpcException(-100, "Unkown block");
+                index = header.Index;
+            }
+            var script_hash = UInt160.Parse(_params[1].AsString());
+            var prefix = Convert.FromBase64String(_params[2].AsString());
+
+            if (StateStore.Singleton.LocalRootIndex < index)
+                throw new RpcException(-100, "Index exceed current index");
+            if (!Settings.Default.FullState && index < StateStore.Singleton.LocalRootIndex)
+            {
+                throw new RpcException(-100, "Old state not supported");
+            }
+            using var snapshot = StateStore.Singleton.GetSnapshot();
+            var state_root = snapshot.GetStateRoot(index);
+            if (state_root is null) throw new InvalidOperationException("state root not found");
+            using var store = StateStore.Singleton.GetStoreSnapshot();
+            trie = new MPTTrie<StorageKey, StorageItem>(store, state_root.RootHash);
+
+            var contract = GetHistoricalContractState(trie, script_hash);
+            if (contract is null) throw new RpcException(-100, "Unknown contract");
+            skey = new()
+            {
+                Id = contract.Id,
+                Key = prefix,
+            };
+        }
+
+        [RpcMethod]
+        public JObject FindState(JArray _params)
+        {
+            PrepareStateParams(_params, out var trie, out var skey);
+            int pageNumber = 0;
+            if (3 < _params.Count)
+                pageNumber = int.Parse(_params[3].AsString());
+            if (Settings.Default.MaxPageNumber <= pageNumber)
```

Comment:
What is the rationale for specifying a max page number?

Reply:
User can use Empty as prefix and scan all storages, Considering there is no memory protection mechanism in MPT impl, it's dangerous to resolve all branches.

Reply:
IIUC it's contract-bound anyway, although NEO or GAS contracts can of course have a lot of data. At the same time, walking through whole NEO or GAS storage doesn't mean storing whole MPT for them in memory, there is no need to (reminds me of nspcc-dev/neo-go#2019 and the way it throws away from memory things it doesn't need).

But it can be improved even further if we're to add `from` path into request, this way you can start resolving `from` (not iterating whole `prefix` until you end up on a requested page) and then walk the number of elements needed and return them. The next request will have new `from`.

Reply:
Use `from` is better.

Reply:
This technically also allows for user-defined paging (specifying number of results to return).

Reply:
Updated.

Reply:
This is a nice improvement. Great suggestion @roman-khimov